### PR TITLE
jest: Update comparison methods to support BigInt

### PR DESCRIPTION
**Pull request by D4nte**
As per https://github.com/facebook/jest/pull/8382

TypeScript 3.2+ supports BigInt with esnext target.
TypeScript 3.8+ supports it with es2020 target.

dtslint is very particular on how the `typesVersions`
should be handled, hence the code duplication.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/jest/pull/8382
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by D4nte** [comment]
How can I see if any of the code maintainers have been notified?

**Review by wsmd — CHANGES_REQUESTED** [review]
Thanks for the PR, @D4nte!

> dtslint is very particular on how the typesVersions
> should be handled, hence the code duplication.

The code duplication isn't ideal since we would need to maintain 3 files at the same time moving forward (i.e. any change to be made in the future, we have to ensure that it's reflected in all copies).

Have you considered referencing the main types in the version-scoped files (as opposed to copying them) and adding only the specific overrides needed? It seems that how it's implemented in [`@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/e0465ecda7451878e007c9c153fc6c4b89c147b5/types/node) (I see they're using `// tslint:disable-next-line:no-bad-reference` in some places).

**Comment by D4nte** [comment]
> Thanks for the PR, @D4nte!
> 
> > dtslint is very particular on how the typesVersions
> > should be handled, hence the code duplication.
> 
> The code duplication isn't ideal since we would need to maintain 3 files at the same time moving forward (i.e. any change to be made in the future, we have to ensure that it's reflected in all copies).
> 
> Have you considered referencing the main types in the version-scoped files (as opposed to copying them) and adding only the specific overrides needed? It seems that how it's implemented in [`@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/e0465ecda7451878e007c9c153fc6c4b89c147b5/types/node) (I see they're using `// tslint:disable-next-line:no-bad-reference` in some places).

I tried but `yarn test` was unhappy. Will try again with some inspiration from `@types/node`.

**Comment by sandersn** [comment]
@D4ante @wsmd is this ready to go? My opinion is that the code-sharing technique is fine -- it technically works even though it's a bit confusing.

**Comment by wsmd** [comment]
@sandersn I'm planning on taking another look later tonight before final approval, but overall this addresses my last suggestion. Sorry for the wait, @D4nte!

**Comment by D4nte** [comment]
Yes, ready to go from my point of view. If not, please do mention what you'd like to see changed :)

File: types/jest/ts3.8/index.d.ts
Comment (wsmd) [inline types/jest/ts3.8/index.d.ts:2]: If all what `types/jest/ts3.8` doing here is referencing the type definitions of `ts3.2`, why wouldn't rely on `>=3.2.0-0` alone? Why do we need to the specify explicitly `>=3.8.0-0`?

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cdfbf3a1dc884d6ea4ca4c80524989f713770d41/types/jest/package.json#L8-L11

Comment (D4nte) [reply types/jest/ts3.8/index.d.ts:2]: This an excellent question with a frustrating answer :)

For `BigInt` to be available, the `target` needs to be:
- `esnext` for TypeScript 3.2 up to <3.8
- `es2020` for 3.8 and above.

I am happy for suggestions but I spent some times on this and could not find a way to work around it (but I am not a TS expert).

Comment (wsmd) [reply types/jest/ts3.8/index.d.ts:2]: I am not sure how this `typesVersions` configuration is related `target` or works around it.

If TypeScript itself supports bigints at a given release, the target version of the project might not support it. At which point, you will get an error related to the configuration itself asking you to change the `target`. You will not get a type error related to jest types.

For 3.2 and up to <3.8, bigints only work under the `esnext` target. However, bigints in 3.8 work under `es2020` as well as `esnext` (bigints were marked as stable for `es2020` in 3.8).

I tested `3.2.4`, `3.8.3` (vscode),  and `3.9.2` with the following alone and I did not encounter any issues.

```json
"typesVersions": { 
   ">=3.2.0-0": { "*": ["ts3.2/*"] } 
}
``` 

All versions worked as expected (but again, `>=3.2.0 <3.8.0` requires `esnext` for this to work but that's not related to the `typesVersions` configuration in this PR).

Comment (D4nte) [reply types/jest/ts3.8/index.d.ts:2]: Ok, I am currently running the test without the `ts3.8` folder. I thought that I just could not make all tests pass but I must say I don't remember now.

Comment (D4nte) [reply types/jest/ts3.8/index.d.ts]: Indeed, I stand corrected, it actually works with `esnext` target for all versions `>=3.2.0-0`. PR has been updated.

**Review by wsmd — APPROVED** [review]
Looks good! Thanks for doing this @D4nte!

**Pull request merged**
